### PR TITLE
Implement C package: Declarations8

### PR DESCRIPTION
**Pull request by knewbury01**
## Description

Decl8

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [x] Queries have been added for the following rules:
     - DCL30-C
 - [ ] Queries have been modified for the following rules:
     - _rule number here_

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [ ] Yes
  - [x] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [ ] Have you considered/identified possible edge cases?
 - [ ] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

**Comment by knewbury01** [comment]
# Notes DCL30-C

I chose to implement a specific subset of cases for this rule, as opposed to the general title/description of the rule. The examples in the wiki only describe the cases that I chose to cover and the tool comparison part of the wiki/logistics details described confirm that handling more general cases is a more difficult problem

### This query was very close to several pre-existing queries...

1) the C++ rule [M7-5-1](https://github.com/github/codeql-coding-standards/blob/main/cpp/autosar/src/rules/M7-5-1/FunctionReturnAutomaticVarCondition.ql) covers one specific part of the cases that I chose to cover (however C does not have references so I opted to write one query to unify the cases and omit the unnecessary reference bits)
2) the out of the box queries - [ReturnStackAllocatedMemory](https://github.com/github/codeql/blob/main/cpp/ql/src/Likely%20Bugs/Memory%20Management/ReturnStackAllocatedMemory.ql) and [StackAddressEscapes](https://github.com/github/codeql/blob/main/cpp/ql/src/Likely%20Bugs/Memory%20Management/StackAddressEscapes.ql) also cover pretty similar scenarios, however as I said, I chose to unify the results into one query and additionally `StackAddressEscapes` has the same FP that my version does so I think maybe the results are not too different?

### possible expansions

if it seems necessary for this query to cover more of the ways that storage durations can conflict or lead to undefined behaviour , possibly [this query](https://github.com/github/codeql-coding-standards/blob/main/c/cert/src/rules/CON34-C/AppropriateThreadObjectStorageDurations.ql) for thread storage durations could be a shared query split for the thread case specifically

**Comment by lcartey** [comment]
There is actually an even closer shared query we've written: [DoNotCopyAddressOfAutoStorageObjectToOtherObject.qll](https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/rules/donotcopyaddressofautostorageobjecttootherobject/DoNotCopyAddressOfAutoStorageObjectToOtherObject.qll)

This is used in:
 * `M7-5-2` `AssignmentOfEscapingAutoStorage.ql`
 * `Rule 18.6` `AutomaticStorageObjectAddressCopiedToOtherObject.ql`

I think the query you've written covers essentially the same topics, so I think I would recommend replacing your query with an import query for the same library. The shared query could be extended if it doesn't cover all the cases (return statements?).

**Comment by knewbury01** [comment]
> There is actually an even closer shared query we've written: [DoNotCopyAddressOfAutoStorageObjectToOtherObject.qll](https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/rules/donotcopyaddressofautostorageobjecttootherobject/DoNotCopyAddressOfAutoStorageObjectToOtherObject.qll)
> 
> This is used in:
> 
> * `M7-5-2` `AssignmentOfEscapingAutoStorage.ql`
> * `Rule 18.6` `AutomaticStorageObjectAddressCopiedToOtherObject.ql`
> 
> I think the query you've written covers essentially the same topics, so I think I would recommend replacing your query with an import query for the same library. The shared query could be extended if it doesn't cover all the cases (return statements?).

@lcartey, thanks for the review! I took a look,

I think DCL30-C == M7-5-1 + M7-5-2 (this one is actually using the out of the box equivalent for stack addresses that I found, but is the refined for our purposes version, so thank you for the much better find!)

I have now split and use the shared lib for the M7-5-2 part

however now that I look at the `M7-5-1` implementation , I still think that maybe my version of it makes more sense for C (omitting the references stuff and using dataflow instead of not), unless you think that both need an update?

File: c/cert/src/rules/DCL30-C/AppropriateStorageDurationsFunctionReturn.ql [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ *       external/cert/obligation/rule
+ */
+
+import cpp
+import codingstandards.c.cert
+import codingstandards.cpp.FunctionParameter
+import semmle.code.cpp.dataflow.DataFlow
+
+class Source extends StackVariable {
+  Source() { not this instanceof FunctionParameter }
```
Comment (lcartey) [inline c/cert/src/rules/DCL30-C/AppropriateStorageDurationsFunctionReturn.ql:20]: I don't think `FunctionParameter` is needed here - you want to exclude all parameters, not just those from definitions or instantiated templates.
```suggestion
  Source() { not this instanceof Parameter }
```

File: c/cert/src/rules/DCL30-C/AppropriateStorageDurationsFunctionReturn.ql [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+abstract class Sink extends DataFlow::Node { }
+
+class FunctionSink extends Sink {
```
Comment (lcartey) [inline c/cert/src/rules/DCL30-C/AppropriateStorageDurationsFunctionReturn.ql:25]: Combine `FunctionSink` and `Sink` - you don't need the abstract class, as you only have one sub-class.

File: c/cert/src/rules/DCL30-C/AppropriateStorageDurationsFunctionReturn.ql [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  FunctionSink() {
+    //output parameter
+    exists(FunctionParameter f |
```
Comment (lcartey) [inline c/cert/src/rules/DCL30-C/AppropriateStorageDurationsFunctionReturn.ql:28]: Again, we don't need `FunctionParameter` here.
```suggestion
    exists(Parameter f |
```

**Review by lcartey — APPROVED** [review]
Just some minor stylistic comments, otherwise looks good to me!

File: c/cert/src/rules/DCL30-C/AppropriateStorageDurationsFunctionReturn.ql (hunk 1 above)
Comment (knewbury01) [reply c/cert/src/rules/DCL30-C/AppropriateStorageDurationsFunctionReturn.ql:20]: agree, addressed!

File: c/cert/src/rules/DCL30-C/AppropriateStorageDurationsFunctionReturn.ql (hunk 2 above)
Comment (knewbury01) [reply c/cert/src/rules/DCL30-C/AppropriateStorageDurationsFunctionReturn.ql:25]: right, I think that was leftover maybe from the previous unified query, good catch!

File: c/cert/src/rules/DCL30-C/AppropriateStorageDurationsFunctionReturn.ql (hunk 3 above)
Comment (knewbury01) [reply c/cert/src/rules/DCL30-C/AppropriateStorageDurationsFunctionReturn.ql:28]: agree, addressed!

**Pull request merged**